_posts/2019-09-03-faster-strings-cpp-parquet.md

```diff
@@ -0,0 +1,233 @@
+  `arrow::DictionaryArray` so that reading such data back is more efficient
+
+All of these things together have produced some excellent performance results
+that we will detail below.
+
+# Pushing Arrow columnar read and write lower in the Parquet stack
+
+The other class of optimizations we implemented was removing an abstraction
+layer between the low-level Parquet column data encoder and decoder classes and
+the Arrow columnar data structures. This involves:
+
+* Adding `ColumnWriter::WriteArrow` and `Encoder::Put` methods that accept
+  `arrow::Array` objects directly
+* Adding `ByteArrayDecoder::DecodeArrow` method to decode binary data directly
+  into an `arrow::BinaryBuilder`.
+
+While the performance improvements from this work are less dramatic than for
+dictionary-encoded data, they are still meaningful in real-world applications.
+
+# Performance Benchmarks
+
+We run some benchmarks comparing Arrow 0.11.1 with the current master
+branch. We construct two kinds of Arrow tables with 10 columns each:
+
+* "Low cardinality" and "high cardinality" variants. The "low cardinality" case
+  has 1,000 unique string values of 32-bytes each. The "high cardinality" has
+  100,000 unique values
+* "Dense" (non-dictionary) and "Dictionary" variants
+
+[Click here is the full benchmark script.][11]
+
+We show both single-threaded and multithreaded read performance. The test
+machine is an Intel i9-9960X using gcc 8.3.0 (on Ubuntu 18.04) with 16 physical
+cores and 32 virtual cores.
+
+First, the writing benchmarks:
+
+<div align="center">
+<img src="{{ site.base-url }}/img/20190903_parquet_write_perf.png"
+     alt="Parquet write benchmarks"
+     width="80%" class="img-responsive">
+</div>
+
+Here we note that writing `arrow::DictionaryArray` is dramatically faster due
+to the optimizations described above. We have achieved a small improvement in
+writing dense (non-dictionary) binary arrays.
+
+Then, the reading benchmarks:
+
+<div align="center">
+<img src="{{ site.base-url }}/img/20190903_parquet_read_perf.png"
+     alt="Parquet read benchmarks"
+     width="80%" class="img-responsive">
+</div>
+
+Here, similarly reading `DictionaryArray` directly is many times faster.
+
+In these benchmarks we note that reading the dense binary data is slower in
+master branch than in version 0.11.0, so we will need to do some profiling and
+see what we can do to bring read performance back inline. Optimizing the dense
+read path has not been too much of a priority relative to the dictionary read
+path in this work.
+
+# Memory Use Improvements
+
+In addition to faster performance, reading columns as dictionary-encoded can
+yield significantly less memory use.
+
+In the `dict-random` case above, we found that the master branch uses 405 MB of
+RAM at peak while loading a 152 MB dataset. In v0.11.1, loading the same
+Parquet file without the accelerated dictionary support uses 1.94 GB of peak
+memory while the resulting non-dictionary table occupies 1.01 GB.
```

Comment:
Isn't there an even bigger difference when compared with 0.14?

Reply:
I'll mention that 0.14.x has a nasty bug in it